### PR TITLE
Refactoring Steerable Behaviours to OO

### PR DESCRIPTION
**Pull request by stellarxo**
breaking down steerabletest

### CODE REVIEW

**Comment by stellarxo** [comment]
so right now, we're basically leaving steerable as is 
and steering test will be broken down

each lightsource instance will be assigned its own id, which is going to be used in a priority container of sorts (right now, it's a dictionary, in which the ID is the key and the value is the strategy)

FishA: wander (priority 0), flee (1)
FishB: wander (0), seek(1), flee(2)
FishC: wander(0), seek(1)

Basically each fish's container will have wander, which is executed until something with higher priority is put into the container (i.e. if FishA detects another fish in its trigger collider, then it will queue up 'flee' and then execute that first)

**Comment by stellarxo** [comment]
Everything of the same priority number will have to be executed at the same time. For example, if fish a is trying to flee from 3 game objects at the same time, it should be able to calculate the correct vector and run away in a direction away from all 3. 

Also to consider, things like cohesion will be running at the same time as wander for fish in schools.
Also, obstacle avoidance should always be running..

**Comment by stellarxo** [comment]
@johnny-la

**Comment by johnny-la** [comment]
Definitely sounds like a smart way of building an AI controller

**Comment by johnny-la** [comment]
I've had a thorough read-through of the code, and you've done a great job overall. I have a set of comments to take into consideration @stellarxo:

### Comments/TODOs
* Is there a reason why _actionDictionary_ and _stoppingCondition_ are public? It seems like they should only be accessible from AbstractFish and its subclasses.
* The _player_, _target_ and _other_ variables have no access modifier. It's considered best practice to always include one. From the looks of those variables, it seems like a _private_ modifier would work best.
* In its current state, _isProximateToNPC_ might hold false information. Consider the case where I have two NPCs that enter the fish's detection volume. If one of them exits and the other stays, OnTriggerExit() will be called and _isProximateToNPC_ will be **false** despite that there's another NPC still in the fish's detection volume. I consider introducing a `numberCloseNPC:int` private variable to keep track of the _amount_ of NPCs close to the fish. Then, an `IsProximateToNPC():bool` function could return true iff `numberCloseNPC > 0`
* The indentation for some lines are off. It's likely because you've done some copy-pasting from my `Steerable` class, which does not follow the same conventions as this project.
* We could have a set of "always running" steering behaviours such as cohesion, alignment and obstacle avoidance stored in a `List<NPCActionable>`.
